### PR TITLE
[training migration] Add serialization features to config container

### PR DESCRIPTION
**Pull request by maanug-nv**
# What does this PR do ?
Enables serializing+deserializing config containers to/from yaml or dictionary.

Also includes an allowlist for instantiating from yaml/dict for security purposes.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/training/argument_utils.py [hunk 1]
```diff
@@ -248,3 +261,58 @@ def _get_field_docstrings(self, src_cfg_class: type) -> dict[str, str]:
...
+    for f in fields(cls):
+        if hasattr(args, f.name):
+            kwargs[f.name] = getattr(args, f.name)
+
+    if return_instance:
+        return cls(**kwargs)
+    else:
+        return kwargs
+
+def pretrain_container_from_args(args: Namespace) -> PretrainConfigContainer:
```
Comment (yobibyte) [inline megatron/training/argument_utils.py]: I would rename this to pretrain_cfg_container_from_args. Alternatively, we can make this a class method of PretrainConfigContainer and just name it `from_args`

Comment (maanug-nv) [reply megatron/training/argument_utils.py]: I'd like to move all arguments related stuff outside of `megatron/training` eventually, including everything in `argument_utils.py` . The training loop is packaged with MCore right now, and ideally that code should be interface agnostic (i.e. offers dataclasses and users can decide how to fill those out).

I'll rename it to `pretrain_cfg_container_from_args`.

File: megatron/training/config/container.py [hunk 2]
```diff
@@ -1,13 +1,218 @@
...
+        Create a config container from a YAML file.
+
+        Args:
+            yaml_path: Path to the YAML file
+            mode: Serialization mode (strict or lenient)
+
+        Returns:
+            A new instance of this class initialized with the YAML file values
+        """
+        if MultiStorageClientFeature.is_enabled():
```
Comment (jaredcasper) [inline megatron/training/config/container.py:83]: This pattern keeps showing up, here and other places... any way to clean this up? I thought MSC was supposed to be mostly transparent in the code?

File: megatron/training/config/instantiate_utils.py [hunk 3]
```diff
@@ -0,0 +1,552 @@
+# Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
```
Comment (jaredcasper) [inline megatron/training/config/instantiate_utils.py]: Wrong year.

File: megatron/training/config/instantiate_utils.py [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+    """
+
+    def __init__(self) -> None:
+        self._allowed_prefixes: list[str] = list(_DEFAULT_ALLOWED_PREFIXES)
+        self._allowed_exact: set[str] = set(_DEFAULT_ALLOWED_EXACT)
+        self._enabled: bool = True
+        self._check_env_override()
+
+    def _check_env_override(self) -> None:
+        if os.environ.get("MEGATRON_ALLOW_ALL_TARGETS", "").strip() == "1":
```
Comment (jaredcasper) [inline megatron/training/config/instantiate_utils.py]: Is there any other way than an environment variable?? Why do we even need this option?

Comment (maanug-nv) [reply megatron/training/config/instantiate_utils.py]: i'm fine to drop this method of control completely. i guess the env var allows you to disable the allowlist checking without modifying your python script, thought some user doing heavy customization might want this when loading their own  classes from a yaml.
but looking at it again, it's just as easy to import this and do `target_allowlist.disable()`.

File: megatron/training/config/container.py (hunk 2 above)
Comment (maanug-nv) [reply megatron/training/config/container.py:83]: naive solution would be to create a bunch of helper functions that create handle different file operations just like is being done here. eg:
```py
def msc_exists(*args, **kwargs):
    if MultiStorageClientFeature.is_enabled():
        msc = MultiStorageClientFeature.import_package()
        return msc.os.path.exists(*args, **kwargs)
    else:
        return os.path.exists(*args, **kwargs)
```

but that could be a lot of functions and would require updating.

it looks this is all the modules that we use from msc:
```
msc.open
msc.os
msc.Path
msc.torch
```

so another solution could be like this:
```py
class MaybeMSC:
    def __init__(self):
        if MultiStorageClientFeature.is_enabled():
            msc = MultiStorageClientFeature.import_package()
            self.open = msc.open
            self.os = msc.os
            self.Path = msc.Path
            self.torch = msc.torch
        else:
            import os, torch
            from pathlib import Path
            self.open = open
            self.os = os
            self.Path = Path
            self.torch = torch
maybe_msc = MaybeMSC()
...
maybe_msc.os.path.exists(yaml_path) # will automatically use the correct module        
```

Comment (maanug-nv) [reply megatron/training/config/container.py:83]: would prefer to do that in a separate pr. if you are ok with one of these approaches, ill create an issue

**Review by jaredcasper — APPROVED** [review]

File: megatron/training/config/container.py (hunk 2 above)
Comment (maanug-nv) [reply megatron/training/config/container.py:83]: created #4507 for this

**Review by cuichenx — APPROVED** [review]
LGTM
